modules/apps/foundation/captcha/captcha-api/src/main/java/com/liferay/captcha/configuration/CaptchaConfigurationModelListener.java

```diff
@@ -0,0 +1,165 @@
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.LocaleThreadLocal;
+import com.liferay.portal.kernel.util.PropsKeys;
+import com.liferay.portal.kernel.util.ResourceBundleUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.security.lang.DoPrivilegedBean;
+import com.liferay.portal.util.PrefsPropsUtil;
+
+import java.util.Dictionary;
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+import javax.portlet.PortletPreferences;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Pei-Jung Lan
+ */
+@Component(
+	immediate = true,
+	property = {
+		"model.class.name=com.liferay.captcha.configuration.CaptchaConfiguration"
+	},
+	service = ConfigurationModelListener.class
+)
+public class CaptchaConfigurationModelListener
+	implements ConfigurationModelListener {
+
+	public void onBeforeSave(String pid, Dictionary<String, Object> properties)
+		throws ConfigurationModelListenerException {
+
+		try {
+			updateCaptcha(properties);
+		}
+		catch (CaptchaConfigurationException cce) {
+			throw new ConfigurationModelListenerException(
+				cce.getMessage(), CaptchaConfiguration.class,
+				CaptchaConfigurationModelListener.class, properties);
+		}
+		catch (Exception e) {
+			_log.error(e, e);
+
+			throw new ConfigurationModelListenerException(
+				ResourceBundleUtil.getString(
+					getResourceBundle(),
+					"an-error-occurred-while-updating-the-captcha-engine"),
+				CaptchaConfiguration.class,
+				CaptchaConfigurationModelListener.class, properties);
+		}
+	}
+
+	protected ResourceBundle getResourceBundle() {
+		if (_resourceBundle == null) {
+			Locale locale = LocaleThreadLocal.getThemeDisplayLocale();
+
+			return ResourceBundleUtil.getBundle(
+				"content.Language", locale, getClass());
+		}
+
+		return _resourceBundle;
+	}
+
+	protected void updateCaptcha(Dictionary<String, Object> properties)
+		throws Exception {
+
+		String captchaEngine = (String)properties.get("captchaEngine");
+		String reCaptchaPublicKey = (String)properties.get(
+			"reCaptchaPublicKey");
+		String reCaptchaPrivateKey = (String)properties.get(
+			"reCaptchaPrivateKey");
+
+		Captcha captcha = null;
+
+		if (captchaEngine.equals(SimpleCaptchaImpl.class.getName())) {
+			captcha = new SimpleCaptchaImpl();
+		}
+		else {
+			captcha = new ReCaptchaImpl();
+
+			validateReCaptchaKeys(reCaptchaPublicKey, reCaptchaPrivateKey);
+		}
+
+		PortletPreferences portletPreferences = PrefsPropsUtil.getPreferences();
+
+		Class<?> clazz = captcha.getClass();
+
+		portletPreferences.setValue(
+			PropsKeys.CAPTCHA_ENGINE_IMPL, clazz.getName());
+
+		portletPreferences.setValue(
+			PropsKeys.CAPTCHA_ENGINE_RECAPTCHA_KEY_PRIVATE,
+			reCaptchaPrivateKey);
+		portletPreferences.setValue(
+			PropsKeys.CAPTCHA_ENGINE_RECAPTCHA_KEY_PUBLIC, reCaptchaPublicKey);
+
+		portletPreferences.store();
```

Comment:
Yeah, we should no longer need to do this, since we are moving the configuration to Config Admin. What we will need is an upgrade process.

BTW, there are some cases in which we will still need to preserve portletPreferences as a persistence (for now) and that is when we need different configuration per company (as what happens with the entries configured through Instance Settings). But I don't think that is the case here, right?